packages/components/src/components/SquiggleViewer/SqViewNode.tsx

```diff
@@ -0,0 +1,166 @@
+import { SqValue, SqValuePath } from "@quri/squiggle-lang";
+
+import { getChildrenValues, TraverseCalculatorEdge } from "./utils.js";
+
+//We might want to bring this into the SquiggleLang library. The ``traverseCalculatorEdge`` part is awkward though.
+class SqValueNode {
+  constructor(
+    public root: SqValue,
+    public path: SqValuePath,
+    public traverseCalculatorEdge: TraverseCalculatorEdge
+  ) {}
+
+  uid() {
+    return this.path.uid();
+  }
+
+  isEqual = (other: SqValueNode): boolean => {
+    return this.uid() === other.uid();
+  };
+
+  sqValue(): SqValue | undefined {
+    return this.root.getSubvalueByPath(this.path, this.traverseCalculatorEdge);
+  }
+
+  parent() {
+    const parentPath = this.path.parent();
+    return parentPath
+      ? new SqValueNode(this.root, parentPath, this.traverseCalculatorEdge)
+      : undefined;
+  }
+
+  children() {
+    const value = this.sqValue();
+    if (!value) {
+      return [];
+    }
+    return getChildrenValues(value)
+      .map((childValue) => {
+        const path = childValue.context?.path;
+        return path
+          ? new SqValueNode(this.root, path, this.traverseCalculatorEdge)
+          : undefined;
+      })
+      .filter((a) => a !== undefined) as SqValueNode[];
+  }
+
+  lastChild(): SqValueNode | undefined {
+    return this.children().at(-1);
+  }
+
+  siblings() {
+    return this.parent()?.children() ?? [];
+  }
+
+  prevSibling() {
+    const index = this.getParentIndex();
+    const isRootOrError = index === -1;
+    const isFirstSibling = index === 0;
+    if (isRootOrError || isFirstSibling) {
+      return undefined;
+    }
+    return this.siblings()[index - 1];
+  }
+
+  nextSibling() {
+    const index = this.getParentIndex();
+    const isRootOrError = index === -1;
+    const isLastSibling = index === this.siblings().length - 1;
+    if (isRootOrError || isLastSibling) {
+      return undefined;
+    }
+    return this.siblings()[index + 1];
```

Comment:
Probably not a problem in practice, but I notice that this method constructs siblings 3 times; each time it goes all the way down to `SqArray.getValues()` or `SqDict.entries()`, calls `wrapValue`, generates new paths, etc.

Reply:
Yea, I didn't worry about this much at this point. We could improve later with caching or something.